File: backend/app/models/requirement.py
from sqlmodel import Field, SQLModel, Relationship
from typing import Optional, List
from datetime import datetime
from .article import ArticleCreate
from decimal import Decimal

class RequirementBase(SQLModel):
    project_id: Optional[int] = Field(foreign_key="project.id", default=None)
    request_date: datetime = Field(default_factory=datetime.utcnow)
    requested_by: Optional[int] = Field(foreign_key="user.id", default=None)
    state_id: int = Field(foreign_key="requirementstate.id")
    closing_date: Optional[datetime] = None

class Requirement(RequirementBase, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    
    # Relationships
    project: Optional["Project"] = Relationship(back_populates="requirements")
    requester: Optional["User"] = Relationship(back_populates="requested_requirements")
    state: "RequirementState" = Relationship(back_populates="requirements")
    articles: List["Article"] = Relationship(back_populates="requirement")

class RequirementCreate(RequirementBase):
    pass

class RequirementResponse(RequirementBase):
    id: int

class RequirementWithArticlesCreate(SQLModel):
    """Model for creating a requirement with its articles"""
    requirement: RequirementCreate
    articles: List["ArticleCreateWithoutRequirement"]

class ArticleCreateWithoutRequirement(SQLModel):
    """Model for creating an article without requirement_id"""
    quantity: Decimal
    unit: str
    brand: str
    model: str
    dimensions: str
    state_id: int = Field(foreign_key="articlestate.id")
    notes: Optional[str] = None 

File: frontend/src/types/requirement.ts
export interface Requirement {
  id: number;
  project_id: number | null;
  request_date: string;
  requested_by: number | null;
  state_id: number;
  closing_date: string | null;
  articles?: Article[];
}

export interface Article {
  id: number;
  requirement_id: number | null;
  requirement_consecutive: number | null;
  quantity: string;
  unit: string;
  brand: string;
  model: string;
  dimensions: string;
  state_id: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface RequirementCreate {
  project_id?: number;
  request_date?: string;
  requested_by?: number;
  state_id: number;
  closing_date?: string;
}

export interface RequirementUpdate {
  project_id?: number;
  request_date?: string;
  requested_by?: number;
  state_id?: number;
  closing_date?: string;
}

// Tipo extendido para mostrar información relacionada
export interface RequirementWithDetails extends Requirement {
  project_name?: string;
  requester_name?: string;
  state_name?: string;
} 